File: frontend/src/api/metrics.ts
import { apiGet } from "./client"
import type {
  ActionEffectivenessResponse,
  MetricsDashboardResponse,
  MetricsOverviewResponse,
  OutcomeMetricsResponse,
} from "./types"

export function fetchMetricsOverview(): Promise<MetricsOverviewResponse> {
  return apiGet<MetricsOverviewResponse>("/api/v1/metrics/overview")
}

export function fetchMetricsDashboard(): Promise<MetricsDashboardResponse> {
  return apiGet<MetricsDashboardResponse>("/api/v1/metrics/dashboard")
}

export function fetchOutcomeMetrics(): Promise<OutcomeMetricsResponse> {
  return apiGet<OutcomeMetricsResponse>("/api/v1/metrics/outcomes")
}

export function fetchActionEffectiveness(actionType?: string): Promise<ActionEffectivenessResponse> {
  const query = actionType ? `?action_type=${encodeURIComponent(actionType)}` : ""
  return apiGet<ActionEffectivenessResponse>(`/api/v1/metrics/actions${query}`)
}

File: backend/app/api/routes/metrics.py
from fastapi import APIRouter, Depends, Query

from app.api.deps import get_runtime
from app.schemas.api import MetricsDashboardResponse, MetricsOverviewResponse
from app.schemas.outcomes import ActionEffectivenessResponse, OutcomeMetricsResponse, SegmentIntelligenceResponse
from app.services.dashboard import dashboard_metrics
from app.services.metrics import overview_metrics
from app.services.outcomes import action_effectiveness, aggregate_outcomes, load_outcome_records, segment_intelligence
from app.services.runtime import AppRuntime

router = APIRouter(prefix="/metrics", tags=["metrics"])


@router.get("/overview", response_model=MetricsOverviewResponse)
def metrics_overview(runtime: AppRuntime = Depends(get_runtime)) -> MetricsOverviewResponse:
    return overview_metrics(runtime)


@router.get("/dashboard", response_model=MetricsDashboardResponse)
def metrics_dashboard(runtime: AppRuntime = Depends(get_runtime)) -> MetricsDashboardResponse:
    return dashboard_metrics(runtime)


@router.get("/outcomes", response_model=OutcomeMetricsResponse)
def metrics_outcomes(runtime: AppRuntime = Depends(get_runtime)) -> OutcomeMetricsResponse:
    return aggregate_outcomes(load_outcome_records(runtime))


@router.get("/actions", response_model=ActionEffectivenessResponse)
def metrics_actions(
    action_type: str | None = Query(default=None),
    runtime: AppRuntime = Depends(get_runtime),
) -> ActionEffectivenessResponse:
    return action_effectiveness(load_outcome_records(runtime), action_type=action_type)


@router.get("/segments", response_model=SegmentIntelligenceResponse)
def metrics_segments(runtime: AppRuntime = Depends(get_runtime)) -> SegmentIntelligenceResponse:
    return segment_intelligence(load_outcome_records(runtime))
